Python/eof.py

```diff
@@ -0,0 +1,225 @@
+    Args:
+        ds (xarray.Dataset): datacube
+            containing an ensemble of CanRCM4 models
+    Returns:
+        mask (numpy.ma): masked array
+    """
+    mask = ma.masked_invalid(ds['dv'][0, :, :].values)
+    return mask
+
+
+def mask_flat(ds, mask):
+    """Flattens mask into ensemble shape
+    for compatibility.
+    --------------------------------
+    Args:
+        ds (xarray.Dataset): datacube
+            containing an ensemble of CanRCM4 models
+        mask (numpy.ma): masked array for run in ds
+    Returns:
+        mask (nump.ma): masked array in ensemble shape,
+        ensemble_size x number of grid cells
+    """
+    dv_field = ds['dv'].values
+    n_grid_cells = dv_field.shape[1]*dv_field.shape[2]
+    mask = np.reshape(mask, n_grid_cells)
+
+    return mask
+
+
+def ens_flat(ds):
+    """Flattens data cube into ensemble
+    size x number of cells.
+    --------------------------------
+    Args:
+        ds (xarray Dataset): datacube
+            containing an ensemble of CanRCM4 models
+    Returns:
+        dv_field (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    """
+    check_keys(ds)
+
+    dv_field = ds['dv'].values
+
+    ens_sz = dv_field.shape[0]
+    n_grid_cells = dv_field.shape[1]*dv_field.shape[2]
+
+    dv_field = np.reshape(dv_field, (ens_sz, n_grid_cells))
+
+    return dv_field
+
+
+def rand_sample_index(y_obs, frac):
+    """Randomly sample a range of integers.
+    --------------------------------
+    Args:
+        y_obs (numpy.ndarray): array to be sampled
+        frac (float): fractional size of y_obs
+            to sample
+    Returns:
+        (numpy.ndarray): array containing indices of
+            sampled values
+    """
+    n_grid_cells = y_obs.shape[0]
+    index = np.random.choice(np.arange(n_grid_cells),
+                             int(frac*n_grid_cells))
+    return index
+
+
+def get_obs(ens_arr):
+    """Randomly sample a data cube to generate
+    pseudo observations from ensemble.
+    --------------------------------
+    Args:
+        ds (xarray Dataset): datacube
+            containing an ensemble of
+            CanRCM4 models
+    Returns:
+        y_sample (numpy.ndarray): fraction
+            of random spatially sampled design
+            value grid cells
+        index (numpy.ndarray): index locations
+            of y_sample from original ensemble size
+    """
+    ens_sz = ens_arr.shape[0]
+
+    # randomly select an ensemble member
+    # to sample
+    i = np.random.randint(0, ens_sz-1)
+    y_obs = ens_arr[i, :]
+
+    return y_obs
+
+
+def ens_to_eof(ens_arr):
+    """Perform EOF/PCA dimensionality reduction
+    on ensemble array
+    --------------------------------
+    Args:
+        ens_arr (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    Returns:
+        eofs (numpy.ndarray): transformed EOFs of ens_arr
+    """
+
+    skpca = pca.PCA(0.95)
```

Comment:
Is there any reason why 95% should be configurable? Would we ever want to use anything *other* than 95%?

Reply:
Maybe, but unlikely. It is now set as a parameter with a default value of `explained_variance=0.95` just in case.